lib/handler/configurator/proxy.c

```diff
@@ -265,11 +266,131 @@ static int on_config_ssl_session_cache(h2o_configurator_command_t *cmd, h2o_conf
     return 0;
 }
 
+static int parse_balancer(h2o_configurator_command_t *cmd, yoml_t *node, h2o_socketpool_target_t **targets, size_t target_len)
+{
+    struct proxy_configurator_t *self = (void *)cmd->configurator;
+    yoml_t *lb_type_node = NULL;
+    size_t i;
+    switch (node->type) {
+    case YOML_TYPE_SCALAR:
+        lb_type_node = node;
+        break;
+    case YOML_TYPE_MAPPING:
+        for (i = 0; i < node->data.mapping.size; i++) {
+            yoml_t *key = node->data.mapping.elements[i].key;
+            yoml_t *value = node->data.mapping.elements[i].value;
+            if (key->type != YOML_TYPE_SCALAR) {
+                h2o_configurator_errprintf(cmd, key, "key must be a scalar");
+                return -1;
+            }
+            if (strcasecmp(key->data.scalar, "type") == 0) {
+                if (value->type != YOML_TYPE_SCALAR) {
+                    h2o_configurator_errprintf(cmd, value, "value must be a scalar");
+                    return -1;
+                }
+                lb_type_node = value;
+                break;
+            }
+        }
+        if (lb_type_node == NULL) {
+            h2o_configurator_errprintf(cmd, node, "`type` must exist when proxy.reverse.balancer configured with a mapping");
+            return -1;
+        }
+        break;
+    default:
+        h2o_configurator_errprintf(cmd, node, "proxy.reverse.balancer must be either a scalar or a mapping");
+        return -1;
+    }
+
+    if (strcmp(lb_type_node->data.scalar, "round-robin") == 0) {
+        self->vars->conf.balancer = h2o_balancer_create_rr();
```

Comment:
Relating to my previous comment, I believe that the selected balancer cannot be stored as a value of the context considering the fact that it is an argument to `proxy.reverse.url`, instead of being a separate configuration directive.

Consider the case where multiple mappings are made, in which only the first one has a `balancer` attribute.
```
paths:
  '/foo':
    proxy.reverse.url:
      backends: ...
      balancer: least-conn
  '/foo':
    proxy.reverse.url:
      backends: ...
```

The mapping for `/foo` will incorrectly use an instance of a least-conn balancer (shouldn't it be round-robin?) which is shared with the instance associated to `/abc` (which is an error).